So the TCK CI build fails with:
> [ERROR]   AsyncMethodTest>Arquillian.run:138->testAsyncInvocationInterceptorProvider:234 expected [21] but found [null]
(Quarkus - TCK - MicroProfile REST Client)

Which is really weird because this branch does not contain this TCK runner. @cescoffier @geoand any ideas?

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ClientProxyGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ClientProxyGenerator.java
@@ -54,6 +54,9 @@ public class ClientProxyGenerator extends AbstractGenerator {
 
     static final String CLIENT_PROXY_SUFFIX = "_ClientProxy";
 
+    static final String DELEGATE_METHOD_NAME = "arc$delegate";
+    static final String GET_CONTEXTUAL_INSTANCE_METHOD_NAME = "getContextualInstance";
+
     private final Predicate<DotName> applicationClassPredicate;
 
     public ClientProxyGenerator(Predicate<DotName> applicationClassPredicate) {
@@ -103,7 +106,6 @@ public class ClientProxyGenerator extends AbstractGenerator {
         for (MethodInfo method : getDelegatingMethods(bean)) {
 
             MethodDescriptor originalMethodDescriptor = MethodDescriptor.of(method);
-
             MethodCreator forward = clientProxy.getMethodCreator(originalMethodDescriptor);
 
             // Exceptions
@@ -116,9 +118,27 @@ public class ClientProxyGenerator extends AbstractGenerator {
                 params[i] = forward.getMethodParam(i);
             }
 
+            if (!superClass.equals(Object.class.getName())) {
+                // Skip delegation if proxy is not constructed yet
+                // This check is unnecessary for producers that return an interface
+                // if(!this.bean == null) return super.foo()
+                BytecodeCreator notConstructed = forward
+                        .ifNull(forward.readInstanceField(beanField.getFieldDescriptor(), forward.getThis())).trueBranch();
+                if (Modifier.isAbstract(method.flags())) {
+                    notConstructed.throwException(IllegalStateException.class, "Cannot delegate to an abstract method");
+                } else {
+                    MethodDescriptor superDescriptor = MethodDescriptor.ofMethod(superClass, method.name(),
+                            method.returnType().name().toString(),
+                            method.parameters().stream().map(p -> p.name().toString()).toArray());
+                    notConstructed.returnValue(
+                            notConstructed.invokeSpecialMethod(superDescriptor, notConstructed.getThis(), params));
+                }
+            }
+
             ResultHandle delegate = forward
                     .invokeVirtualMethod(
-                            MethodDescriptor.ofMethod(generatedName, "delegate", DescriptorUtils.typeToString(providerType)),
+                            MethodDescriptor.ofMethod(generatedName, DELEGATE_METHOD_NAME,
+                                    DescriptorUtils.typeToString(providerType)),
                             forward.getThis());
             ResultHandle ret;
 
@@ -170,7 +190,8 @@ public class ClientProxyGenerator extends AbstractGenerator {
     }
 
     void implementDelegate(ClassCreator clientProxy, String providerTypeName, FieldDescriptor beanField) {
-        MethodCreator creator = clientProxy.getMethodCreator("delegate", providerTypeName).setModifiers(Modifier.PRIVATE);
+        MethodCreator creator = clientProxy.getMethodCreator(DELEGATE_METHOD_NAME, providerTypeName)
+                .setModifiers(Modifier.PRIVATE);
         // Arc.container()
         ResultHandle container = creator.invokeStaticMethod(MethodDescriptors.ARC_CONTAINER);
         // bean.getScope()
@@ -196,10 +217,11 @@ public class ClientProxyGenerator extends AbstractGenerator {
     }
 
     void implementGetContextualInstance(ClassCreator clientProxy, String providerTypeName) {
-        MethodCreator creator = clientProxy.getMethodCreator("getContextualInstance", Object.class)
+        MethodCreator creator = clientProxy.getMethodCreator(GET_CONTEXTUAL_INSTANCE_METHOD_NAME, Object.class)
                 .setModifiers(Modifier.PUBLIC);
         creator.returnValue(
-                creator.invokeVirtualMethod(MethodDescriptor.ofMethod(clientProxy.getClassName(), "delegate", providerTypeName),
+                creator.invokeVirtualMethod(
+                        MethodDescriptor.ofMethod(clientProxy.getClassName(), DELEGATE_METHOD_NAME, providerTypeName),
                         creator.getThis()));
     }
 

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/MethodDescriptors.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/MethodDescriptors.java
@@ -115,7 +115,7 @@ final class MethodDescriptors {
             Class[].class, Object[].class);
 
     static final MethodDescriptor CLIENT_PROXY_GET_CONTEXTUAL_INSTANCE = MethodDescriptor.ofMethod(ClientProxy.class,
-            "getContextualInstance", Object.class);
+            ClientProxyGenerator.GET_CONTEXTUAL_INSTANCE_METHOD_NAME, Object.class);
 
     static final MethodDescriptor INJECTABLE_BEAN_DESTROY = MethodDescriptor.ofMethod(InjectableBean.class, "destroy",
             void.class, Object.class,

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Methods.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Methods.java
@@ -19,7 +19,6 @@ package io.quarkus.arc.processor;
 import java.lang.reflect.Modifier;
 import java.util.ArrayList;
 import java.util.Collection;
-import java.util.Collections;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
@@ -80,14 +79,12 @@ final class Methods {
             for (Type interfaceType : classInfo.interfaceTypes()) {
                 ClassInfo interfaceClassInfo = index.getClassByName(interfaceType.name());
                 if (interfaceClassInfo != null) {
-                    Map<TypeVariable, Type> resolved = Collections.emptyMap();
                     addDelegatingMethods(index, interfaceClassInfo, methods);
                 }
             }
             if (classInfo.superClassType() != null) {
                 ClassInfo superClassInfo = index.getClassByName(classInfo.superName());
                 if (superClassInfo != null) {
-                    Map<TypeVariable, Type> resolved = Collections.emptyMap();
                     addDelegatingMethods(index, superClassInfo, methods);
                 }
             }

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java
@@ -36,6 +36,7 @@ import io.quarkus.gizmo.MethodCreator;
 import io.quarkus.gizmo.MethodDescriptor;
 import io.quarkus.gizmo.ResultHandle;
 import io.quarkus.gizmo.TryBlock;
+import java.lang.reflect.Modifier;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.HashMap;
@@ -262,13 +263,31 @@ public class SubclassGenerator extends AbstractGenerator {
             String providerTypeName, FieldDescriptor interceptorChainsField, FieldDescriptor methodsField,
             InterceptionInfo interceptedMethod) {
 
-        MethodCreator forwardMethod = subclass.getMethodCreator(MethodDescriptor.of(method));
+        MethodDescriptor originalMethodDescriptor = MethodDescriptor.of(method);
+        MethodCreator forwardMethod = subclass.getMethodCreator(originalMethodDescriptor);
 
         // Params
         // Object[] params = new Object[] {p1}
         ResultHandle paramsHandle = forwardMethod.newArray(Object.class, forwardMethod.load(method.parameters().size()));
         for (int i = 0; i < method.parameters().size(); i++) {
             forwardMethod.writeArrayValue(paramsHandle, i, forwardMethod.getMethodParam(i));
+        }
+
+        // if(!this.bean == null) return super.foo()
+        BytecodeCreator notConstructed = forwardMethod
+                .ifNull(forwardMethod.readInstanceField(methodsField, forwardMethod.getThis())).trueBranch();
+        ResultHandle[] params = new ResultHandle[method.parameters().size()];
+        for (int i = 0; i < method.parameters().size(); ++i) {
+            params[i] = notConstructed.getMethodParam(i);
+        }
+        if (Modifier.isAbstract(method.flags())) {
+            notConstructed.throwException(IllegalStateException.class, "Cannot delegate to an abstract method");
+        } else {
+            MethodDescriptor superDescriptor = MethodDescriptor.ofMethod(providerTypeName, method.name(),
+                    method.returnType().name().toString(),
+                    method.parameters().stream().map(p -> p.name().toString()).toArray());
+            notConstructed.returnValue(
+                    notConstructed.invokeSpecialMethod(superDescriptor, notConstructed.getThis(), params));
         }
 
         // Forwarding function

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/clientproxy/constructor/ClientProxyConstructorGuardTest.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/clientproxy/constructor/ClientProxyConstructorGuardTest.java
@@ -1,0 +1,57 @@
+/*
+ * Copyright 2019 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.quarkus.arc.test.clientproxy.constructor;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+import io.quarkus.arc.Arc;
+import io.quarkus.arc.ClientProxy;
+import io.quarkus.arc.test.ArcTestContainer;
+import java.io.IOException;
+import javax.enterprise.context.ApplicationScoped;
+import org.junit.Rule;
+import org.junit.Test;
+
+public class ClientProxyConstructorGuardTest {
+
+    @Rule
+    public ArcTestContainer container = new ArcTestContainer(Moo.class);
+
+    @Test
+    public void testProducer() throws IOException {
+        Moo moo = Arc.container().instance(Moo.class).get();
+        assertTrue(moo instanceof ClientProxy);
+        assertEquals(10, ((Moo) ((ClientProxy) moo).getContextualInstance()).val);
+    }
+
+    @ApplicationScoped
+    static class Moo {
+
+        private int val;
+
+        public Moo() {
+            init();
+        }
+
+        void init() {
+            val = 10;
+        }
+
+    }
+
+}

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/constructor/SubclassConstructorGuardTest.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/constructor/SubclassConstructorGuardTest.java
@@ -1,0 +1,75 @@
+/*
+ * Copyright 2019 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.quarkus.arc.test.interceptors.constructor;
+
+import static org.junit.Assert.assertEquals;
+
+import io.quarkus.arc.Arc;
+import io.quarkus.arc.test.ArcTestContainer;
+import io.quarkus.arc.test.interceptors.Simple;
+import java.io.IOException;
+import javax.annotation.Priority;
+import javax.inject.Singleton;
+import javax.interceptor.AroundInvoke;
+import javax.interceptor.Interceptor;
+import javax.interceptor.InvocationContext;
+import org.junit.Rule;
+import org.junit.Test;
+
+public class SubclassConstructorGuardTest {
+
+    @Rule
+    public ArcTestContainer container = new ArcTestContainer(Simple.class, SimpleBean.class,
+            SimpleInterceptor.class);
+
+    @Test
+    public void testConstructorGuard() throws IOException {
+        assertEquals("foo::bar", Arc.container().instance(SimpleBean.class).get().foo());
+    }
+
+    @Simple
+    @Singleton
+    static class SimpleBean {
+
+        private String val;
+
+        public SimpleBean() {
+            init();
+        }
+
+        void init() {
+            this.val = "bar";
+        }
+
+        String foo() {
+            return val;
+        }
+
+    }
+
+    @Simple
+    @Priority(1)
+    @Interceptor
+    public static class SimpleInterceptor {
+
+        @AroundInvoke
+        Object mySuperCoolAroundInvoke(InvocationContext ctx) throws Exception {
+            return "foo" + "::" + ctx.proceed();
+        }
+    }
+
+}

